README.md
[![Github top language](https://img.shields.io/github/languages/top/Base24/base24-helpers.svg?style=for-the-badge)](../../)
[![Codacy grade](https://img.shields.io/codacy/grade/[codacy-proj-id].svg?style=for-the-badge)](https://www.codacy.com/manual/Base24/base24-helpers)
[![Repository size](https://img.shields.io/github/repo-size/Base24/base24-helpers.svg?style=for-the-badge)](../../)
[![Issues](https://img.shields.io/github/issues/Base24/base24-helpers.svg?style=for-the-badge)](../../issues)
[![License](https://img.shields.io/github/license/Base24/base24-helpers.svg?style=for-the-badge)](/LICENSE.md)
[![Commit activity](https://img.shields.io/github/commit-activity/m/Base24/base24-helpers.svg?style=for-the-badge)](../../commits/master)
[![Last commit](https://img.shields.io/github/last-commit/Base24/base24-helpers.svg?style=for-the-badge)](../../commits/master)

<!-- omit in TOC -->
# base24-helpers

<img src="readme-assets/icons/name.png" alt="Project Icon" width="750">

Helper programs to create templates and schemes.

- [base24tools.py](#base24toolspy)
- [baseCompare.py](#basecomparepy)
- [batchIterm2b24.py](#batchiterm2b24py)
- [iterm2base24.py](#iterm2base24py)
- [windowsTerminal2b24.py](#windowsterminal2b24py)
- [schemeSuggest.py](#schemesuggestpy)
- [templateGen.py](#templategenpy)
- [schemeTableGen.py](#schemetablegenpy)
- [Language information](#language-information)
	- [Built for](#built-for)
- [Install Python on Windows](#install-python-on-windows)
	- [Chocolatey](#chocolatey)
	- [Download](#download)
- [Install Python on Linux](#install-python-on-linux)
	- [Apt](#apt)
- [How to run](#how-to-run)
	- [With VSCode](#with-vscode)
	- [From the Terminal](#from-the-terminal)
- [Download](#download-1)
	- [Clone](#clone)
		- [Using The Command Line](#using-the-command-line)
		- [Using GitHub Desktop](#using-github-desktop)
	- [Download Zip File](#download-zip-file)
- [Community Files](#community-files)
	- [Licence](#licence)
	- [Changelog](#changelog)
	- [Code of Conduct](#code-of-conduct)
	- [Contributing](#contributing)
	- [Security](#security)


## base24tools.py

Lib file used by iterm2base24.py and windowsTerminal2b24.py with the following
function:

```python
def process(base24, base24lookup, sourceThemeDict, bgNumber, fgNumber):
	"""process a base24 dict using a lookup table to translate the source theme
	onto the base24 scheme

	Args:
		base24 (dict): a dictionary representing a base24 scheme
		base24lookup (dict): a dictionary mapping base24 colour ids to the
		respective key in the sourceThemeDict
		sourceThemeDict (dict): a dictionary representing the source theme/
		scheme
		bgNumber (int): lower base0N bound for the background 0-9
		fgNumber (int): upper base0N bound for the foreground 0-9. Must be
		greater than bgNumber

	Returns:
		dict: base24 dict to write to scheme file
	"""
```

## baseCompare.py

Compare base16 and base24 Windows terminal themes quickly

```bash
usage: baseCompare.py [-h] [--all] themes

Compare base16 and base24 windows terminal themes quickly

positional arguments:
  themes      directory containing generated windows terminal themes

optional arguments:
  -h, --help  show this help message and exit
  --all       compare all schemes
```

## batchIterm2b24.py

Convert a directory containing .itermcolors to base24 schemes

```bash
usage: batchIterm2b24.py [-h] itermschemes

Convert .itermcolors to base24 scheme in a directory

positional arguments:
  itermschemes  directory containing itermschemes

optional arguments:
  -h, --help    show this help message and exit
```

## iterm2base24.py

Convert a singular .itermcolors to a base24 scheme

```bash
usage: iterm2base24.py [-h] file

Convert .itermcolors to base24 scheme

positional arguments:
  file        file.itermschemes

optional arguments:
  -h, --help  show this help message and exit
```

## windowsTerminal2b24.py

Convert a profiles.json (such as your own) to a number of base24 schmes

```bash
usage: windowsTerminal2b24.py [-h] file

Convert profiles.json to base24 scheme

positional arguments:
  file        profiles.json

optional arguments:
  -h, --help  show this help message and exit
```

## schemeSuggest.py

Suggest scheme colours based on a base24 theme template and the theme file

```bash
usage: schemeSuggest [-h] [--mode MODE] template theme

Pair colours with base24 colour ids

positional arguments:
  template     relative or abs path to the base24 template
  theme        relative or abs path to the theme file

optional arguments:
  -h, --help   show this help message and exit
  --mode MODE  color format: hex (ff00aa), reversehex (aa00ff), rgb (255,0,170), reversergb (170,0,255), dec (1.0,0,0.666),
               reversedec (0.666,0,1.0)
```

## templateGen.py

Take and existing theme and a base24 scheme file and produces a base24 template
file

```bash
usage: templateGen [-h] [--mode MODE] [--fuzz FUZZ] colour_scheme theme

Generate a base24 template from an existing scheme and a theme file

positional arguments:
  colour_scheme  relative or abs path to the base24 colour scheme
  theme          relative or abs path to the theme file

optional arguments:
  -h, --help     show this help message and exit
  --mode MODE    color format: hex (ff00aa), reversehex (aa00ff), rgb (255,0,170), reversergb (170,0,255), dec (1.0,0,0.666),
                 reversedec (0.666,0,1.0)
  --fuzz FUZZ    find 'close' colours and replace these ((r,g,b)+-fuzz: default=0, max-recommended=5)
```


## schemeTableGen.py

Generate a table to put in a scheme project readme

```bash
usage: schemeTableGen.py [-h] scheme

Generate a table to put in a scheme project readme

positional arguments:
  scheme      base24 scheme file

optional arguments:
  -h, --help  show this help message and exit
```

## Language information
### Built for
This program has been written for Python 3 and has been tested with
Python version 3.8.0 <https://www.python.org/downloads/release/python-380/>.

## Install Python on Windows
### Chocolatey
```powershell
choco install python
```
### Download
To install Python, go to <https://www.python.org/> and download the latest
version.

## Install Python on Linux
### Apt
```bash
sudo apt install python3.8
```

## How to run
### With VSCode
1. Open the .py file in vscode
2. Ensure a python 3.8 interpreter is selected (Ctrl+Shift+P > Python:Select
3. Interpreter > Python 3.8)
4. Run by pressing Ctrl+F5 (if you are prompted to install any modules, accept)
### From the Terminal
```bash
./[file].py
```

## Download
### Clone
#### Using The Command Line
1. Press the Clone or download button in the top right
2. Copy the URL (link)
3. Open the command line and change directory to where you wish to
clone to
4. Type 'git clone' followed by URL in step 2
```bash
$ git clone https://github.com/Base24/base24-helpers
```

More information can be found at
<https://help.github.com/en/articles/cloning-a-repository>

#### Using GitHub Desktop
1. Press the Clone or download button in the top right
2. Click open in desktop
3. Choose the path for where you want and click Clone

More information can be found at
<https://help.github.com/en/desktop/contributing-to-projects/cloning-a-repository-from-github-to-github-desktop>

### Download Zip File

1. Download this GitHub repository
2. Extract the zip archive
3. Copy/ move to the desired location

## Community Files
### Licence
MIT License
Copyright (c) Base24
(See the [LICENSE](/LICENSE.md) for more information.)

### Changelog
See the [Changelog](/CHANGELOG.md) for more information.

### Code of Conduct
In the interest of fostering an open and welcoming environment, we
as contributors and maintainers pledge to make participation in our
project and our community a harassment-free experience for everyone.
Please see the
[Code of Conduct](https://github.com/Base24/.github/blob/master/CODE_OF_CONDUCT.md) for more information.

### Contributing
Contributions are welcome, please see the [Contributing Guidelines](https://github.com/Base24/.github/blob/master/CONTRIBUTING.md) for more information.

### Security
Thank you for improving the security of the project, please see the [Security Policy](https://github.com/Base24/.github/blob/master/SECURITY.md) for more information.


CHANGELOG.md
# Changelog
All major and minor version changes will be documented in this file. Details of
patch-level version changes can be found in [commit messages](../../commits/master).

## 2020.1 - 2020/03/30
- Added more information to readme
- Added `windowsTerminal2b24.py` and moved common code between this and
`iterm2base24` to `base24tool.py`

## 2020 - 2020/03/04
- First release


base24tools.py
"""process a base24 dict using a lookup table to translate the source theme
onto the base24 scheme
"""
import sys
from metprint import (
	LogType,
	Logger
)

def process(base24, base24lookup, sourceThemeDict, bgNumber, fgNumber):
	"""process a base24 dict using a lookup table to translate the source theme
	onto the base24 scheme

	Args:
		base24 (dict): a dictionary representing a base24 scheme
		base24lookup (dict): a dictionary mapping base24 colour ids to the
		respective key in the sourceThemeDict
		sourceThemeDict (dict): a dictionary representing the source theme/
		scheme
		bgNumber (int): lower base0N bound for the background 0-9
		fgNumber (int): upper base0N bound for the foreground 0-9. Must be
		greater than bgNumber

	Returns:
		dict: base24 dict to write to scheme file
	"""
	if fgNumber not in range(0, 10) or bgNumber not in range(0, 10):
		Logger().logPrint("fgNumber and bgNumber must be between 0 and 9", LogType.ERROR)
		sys.exit(1)
	if fgNumber < bgNumber:
		Logger().logPrint("fgNumber must be larger than bgNumber", LogType.ERROR)
		sys.exit(1)
	# Keys apart from base03, base04, base05, base0F, base10, base11 can be
	# taken from iterm
	for key in base24lookup:
		base24[key] = sourceThemeDict[base24lookup[key]]

	# Is it really a base16 theme? Compare red and bright red to find out
	if base24["base08"] == base24["base12"]:
		Logger().logPrint("\"" + base24["scheme"] + "\" is a base16 theme", LogType.WARNING)
		base24["scheme"] = "b16" + base24["scheme"]

	# Keys fg to fg (not including), base0F, base10, base11 can be calculated
	#Background through Foreground
	bg = base24["base0" + str(bgNumber)]
	fg = base24["base0" + str(fgNumber)]
	for index in range(bgNumber + 1, fgNumber):
		mult = index - bgNumber
		baseVal = []
		for section in range(0, 5, 2):
			baseVal.append("{:02x}".format(int(bg[0+section:2+section], 16) + int(
				(int(fg[0+section:2+section], 16) - int(bg[0+section:2+section], 16)) *
				mult / (fgNumber - bgNumber)
				)
			))
		base24["base0"+str(index)] = "".join(baseVal)

	#Darker and darkest backgrounds
	for index in range(10, 12):
		mult = 12 - index
		baseVal = []
		for section in range(0, 5, 2):
			baseVal.append("{:02x}".format(int(int(base24["base00"][0+section:2+section], 16) * mult / 3)
			))
		base24["base"+str(index)] = "".join(baseVal)

	#Dark red (variant 2)
	baseVal = []
	for section in range(0, 5, 2):
		baseVal.append("{:02x}".format(int(int(base24["base08"][0+section:2+section], 16) / 2)
		))
	base24["base0F"] = "".join(baseVal)

	# Have the other colours been calculated correctly? Count the number of
	# 000000s
	listOf000000s = [base24[key] for key in base24 if base24[key] == "000000"]
	if len(listOf000000s) > 2: # 2 seems a sensible threshold
		Logger().logPrint("\"" + base24["scheme"].replace("b16", "") + "\" is probably corrupted",
		LogType.WARNING)
		base24["scheme"] = "warn0s" + base24["scheme"]

	return base24


baseCompare.py
#!/usr/bin/env python3.8
"""Compare base16 and base24 windows terminal themes quickly
"""
import sys
import os
import argparse
import platform
import ctypes
import commentjson
from metprint import (
	LogType,
	Logger
)

def cPrint(colourHex):
	''' Print a hex colour '''
	colour = colourHex[1:]
	print("\033[48;2;"+ ";".join([str(int(colour[0:2], 16)),
	str(int(colour[2:4], 16)), str(int(colour[4:6], 16))]) +
	"m  \033[0m", end="")



def printColours(winTerm):
	''' Print the colours for a theme '''
	scheme = commentjson.loads(open(winTerm).read()[:-2])
	keys = ["background", "black", "brightBlack", "foreground", "white",
	"brightWhite", "red", "yellow", "brightYellow", "green", "cyan", "blue",
	"purple", "brightRed", "brightYellow", "brightGreen", "brightCyan",
	"brightBlue", "brightPurple"]
	for key in keys:
		cPrint(scheme[key])
	print()




def main():
	''' Main entry point for cli '''
	parser = argparse.ArgumentParser(
	description="Compare base16 and base24 windows terminal themes quickly")
	parser.add_argument("themes", action="store",
	help="directory containing generated windows terminal themes")
	parser.add_argument("--all", action="store_true",
	help="compare all schemes")
	args = parser.parse_args()
	# Check for and report level8 errors
	if not os.path.isdir(args.themes):
		Logger().logPrint(args.themes + " is not a valid directory", LogType.ERROR)
		sys.exit(1)

	# Windows has crap colour support to use this
	if platform.system() == "Windows":
		kernel32 = ctypes.windll.kernel32
		kernel32.SetConsoleMode(kernel32.GetStdHandle(-11), 7)


	files = {}
	for theme in os.listdir(args.themes):
		key = theme.replace(".json", "").replace("base16-", "").replace("base24-", "")
		if key in files:
			files[key].append(theme)
		else:
			files[key] = [theme]


	# Check for b16 and b24 variant of a file
	for key in files:
		if len(files[key]) == 2: # Both variants
			Logger().logPrint(key, LogType.HEADER)
			Logger().logPrint("  B     W   R Y   G C B M ", LogType.BOLD)
			if "base16-" in files[key][0]: # First one is base16
				printColours(os.path.join(args.themes, files[key][0]))
				printColours(os.path.join(args.themes, files[key][1]))
			else:
				printColours(os.path.join(args.themes, files[key][1]))
				printColours(os.path.join(args.themes, files[key][0]))
		elif args.all:
			Logger().logPrint(files[key][0].replace(".json", ""), LogType.HEADER)
			Logger().logPrint("  B     W   R Y   G C B M ", LogType.BOLD)
			printColours(os.path.join(args.themes, files[key][0]))




if __name__ == "__main__":
	main()


iterm2base24.py
#!/usr/bin/env python3
"""Convert .itermcolors to base24 scheme
"""

import os
import sys
import argparse
import defusedxml.ElementTree as ET
import yaml
from metprint import (
	LogType,
	Logger
)
import base24tools

def rgb_to_hex(rgb):
	''' Converts rgb to hex '''
	return '%02x%02x%02x' % rgb

def iterm2hex(root):
	"""Get hex codes from iterm xml

	Args:
		root (xmlroot): xmlroot

	Returns:
		dict: iterm keys to hex
	"""
	keys = root.findall("./dict/key")
	dicts = root.findall("./dict/dict")
	iterm = {}
	for i, _key in enumerate(keys):
		keyName = keys[i].text
		r = g = b = None
		for index, item in enumerate(dicts[i]):
			if "Red Component" in item.text:
				r = int(float(dicts[i][index+1].text) * 255.0)
			if "Green Component" in item.text:
				g = int(float(dicts[i][index+1].text) * 255.0)
			if "Blue Component" in item.text:
				b = int(float(dicts[i][index+1].text) * 255.0)
		iterm[keyName] = rgb_to_hex((r, g, b))
	return iterm


def genBase24(filename, iterm):
	"""Generate the base24 json object

	Args:
		filename (str): filename from args
		iterm (dict): iterm keys to hex

	Returns:
		dict: base24 dict to write to scheme file
	"""

	base24 = {"author": "Iterm2B24", "scheme": filename.split(".")[0]}
	base24lookup = {
		"base00": "Background Color",
		"base01": "Ansi 0 Color", #Black
		"base02": "Ansi 8 Color", #Bright black
		"base06": "Ansi 7 Color", #White
		"base07": "Ansi 15 Color", #Bright white
		"base08": "Ansi 1 Color", #Red
		"base09": "Ansi 3 Color", #Yellow
		"base0A": "Ansi 12 Color", #Bright yellow (variant 2)
		"base0B": "Ansi 2 Color", #Green
		"base0C": "Ansi 6 Color", #Cyan
		"base0D": "Ansi 4 Color", #Blue
		"base0E": "Ansi 5 Color", #Purple
		"base12": "Ansi 9 Color", #Bright red
		"base13": "Ansi 11 Color", #Bright yellow
		"base14": "Ansi 10 Color", #Bright green
		"base15": "Ansi 14 Color", #Bright cyan
		"base16": "Ansi 12 Color", #Bright blue
		"base17": "Ansi 13 Color", #Bright purple
	}

	return base24tools.process(base24, base24lookup, iterm, 2, 6)


def main():
	''' Main entry point for cli '''
	parser = argparse.ArgumentParser(
	description="Convert .itermcolors to base24 scheme")
	parser.add_argument("file", action="store",
	help="file.itermschemes")
	args = parser.parse_args()
	# Check for and report level8 errors
	if not os.path.isfile(args.file):
		Logger().logPrint(args.file + " is not a valid file", LogType.ERROR)
		sys.exit(1)

	filename = args.file
	tree = ET.parse(filename)

	base24 = genBase24(filename, iterm2hex(tree.getroot()))

	with open(base24["scheme"]+".yaml", "w") as outfile:
		Logger().logPrint("writing \"" + base24["scheme"] + "\" to file", LogType.SUCCESS)
		yaml.dump(base24, outfile)

if __name__ == '__main__':
	main()


schemeSuggest.py
"""Suggest scheme colours based on a template and a theme file. eg
kde-plasma template and a kde-plasma theme
"""
import argparse
import difflib
import diff_match_patch as dmp_module


def getColourMap(line1, line2):
	diff = difflib.SequenceMatcher(None, line1, line2)
	colourMap = []
	dmp = dmp_module.diff_match_patch()
	dmp.Diff_Timeout = 10
	diff = dmp.diff_main(line1, line2)
	dmp.diff_cleanupSemantic(diff)

	for index in range(len(diff)):
		if diff[index][0] == -1:
			colourMap = (tuple(diff[index][1].split(",")), tuple(diff[index+1][1].split(",")))

	return colourMap


def schemeSuggest(template, theme, mode):
	# Read lines and ignore blank lines (other different lines will cause issues)
	template_lines_d = open(template, "r").readlines()
	template_lines = [template_lines_d[i] for i in range(len(template_lines_d)) if len(template_lines_d[i].strip()) > 0]
	theme_lines_d = open(theme, "r").readlines()
	theme_lines = [theme_lines_d[i] for i in range(len(theme_lines_d)) if len(theme_lines_d[i].strip()) > 0]

	colours = []
	for index in range(len(theme_lines)):
		colours.append(getColourMap(template_lines[index], theme_lines[index]))

	unique = [list(x) for x in set(tuple(x) for x in colours)]
	for element in unique:
		if len(element) > 0:
			try:
				print(element[0][0][2:8], end=": ")
				if "reverse" in mode:
					colour = element[1][::-1]
				else:
					colour = element[1]

				if "rgb" in mode:
					colour = "{:02x}".format(int(colour[0]))+"{:02x}".format(int(colour[1]))+"{:02x}".format(int(colour[2]))
				elif "dec" in mode:
					colour = "{:02x}".format(int(colour[0])*255)+"{:02x}".format(int(colour[1])*255)+"{:02x}".format(int(colour[2])*255)

				print("\033["+ ";".join(["48", "2", str(int(colour[0:2], 16)),str(int(colour[2:4], 16)),str(int(colour[4:6], 16))]) + "m     \033[0m #" + colour)
			except:
				pass


if __name__ == "__main__":
	"""Main method, do argparsing and call schemeSuggest
	"""
	parser = argparse.ArgumentParser("schemeSuggest",
	description="Pair colours with base24 colour ids")
	parser.add_argument("template",
	help="relative or abs path to the base24 template")
	parser.add_argument("theme",
	help="relative or abs path to the theme file")
	parser.add_argument("--mode", action="store", default="hexhash",
	help="""color format: hex (ff00aa), reversehex (aa00ff), rgb (255,0,170),
	reversergb (170,0,255), dec (1.0,0,0.666), reversedec (0.666,0,1.0)""")

	args = parser.parse_args()
	schemeSuggest(args.template, args.theme, args.mode)


batchIterm2b24.py
#!/usr/bin/env python3
"""Convert .itermcolors to base24 scheme in a directory
"""
import sys
import os
import argparse
import yaml
import defusedxml.ElementTree as ET
from metprint import (
	LogType,
	Logger
)
import iterm2base24

def main():
	''' Main entry point for cli '''
	parser = argparse.ArgumentParser(
	description="Convert .itermcolors to base24 scheme in a directory")
	parser.add_argument("itermschemes", action="store",
	help="directory containing itermschemes")
	args = parser.parse_args()
	# Check for and report level8 errors
	if not os.path.isdir(args.itermschemes):
		Logger().logPrint(args.itermschemes + " is not a valid directory", LogType.ERROR)
		sys.exit(1)

	for file in os.listdir(args.itermschemes):
		# Generate scheme from file
		filename = os.path.join(args.itermschemes, file)
		tree = ET.parse(filename)
		base24 = iterm2base24.genBase24(file, iterm2base24.iterm2hex(tree.getroot()))

		# Write schemes
		directory = "batchSchemes"
		if not os.path.exists(directory):
			os.makedirs(directory)
		with open(os.path.join(directory, base24["scheme"]+".yaml"), "w") as outfile:
			Logger().logPrint("writing \"" + base24["scheme"] + "\" to file", LogType.SUCCESS)
			yaml.dump(base24, outfile)

if __name__ == "__main__":
	main()


schemeTableGen.py
"""Generate a table to put in a scheme project readme
"""
import os
import sys
import argparse
import yaml
import pystache


template = '''
# project-name

Base24 scheme for {{scheme-name}}

|baseNN|Colour|
|---|---|
|base00|![#](https://placehold.it/25/{{base00-hex}}/000000?text=+)
|base01|![#](https://placehold.it/25/{{base01-hex}}/000000?text=+)
|base02|![#](https://placehold.it/25/{{base02-hex}}/000000?text=+)
|base03|![#](https://placehold.it/25/{{base03-hex}}/000000?text=+)
|base04|![#](https://placehold.it/25/{{base04-hex}}/000000?text=+)
|base05|![#](https://placehold.it/25/{{base05-hex}}/000000?text=+)
|base06|![#](https://placehold.it/25/{{base06-hex}}/000000?text=+)
|base07|![#](https://placehold.it/25/{{base07-hex}}/000000?text=+)
|base08|![#](https://placehold.it/25/{{base08-hex}}/000000?text=+)
|base09|![#](https://placehold.it/25/{{base09-hex}}/000000?text=+)
|base0A|![#](https://placehold.it/25/{{base0A-hex}}/000000?text=+)
|base0B|![#](https://placehold.it/25/{{base0B-hex}}/000000?text=+)
|base0C|![#](https://placehold.it/25/{{base0C-hex}}/000000?text=+)
|base0D|![#](https://placehold.it/25/{{base0D-hex}}/000000?text=+)
|base0E|![#](https://placehold.it/25/{{base0E-hex}}/000000?text=+)
|base0F|![#](https://placehold.it/25/{{base0F-hex}}/000000?text=+)
|base10|![#](https://placehold.it/25/{{base10-hex}}/000000?text=+)
|base11|![#](https://placehold.it/25/{{base11-hex}}/000000?text=+)
|base12|![#](https://placehold.it/25/{{base12-hex}}/000000?text=+)
|base13|![#](https://placehold.it/25/{{base13-hex}}/000000?text=+)
|base14|![#](https://placehold.it/25/{{base14-hex}}/000000?text=+)
|base15|![#](https://placehold.it/25/{{base15-hex}}/000000?text=+)
|base16|![#](https://placehold.it/25/{{base16-hex}}/000000?text=+)
|base17|![#](https://placehold.it/25/{{base17-hex}}/000000?text=+)

proj-down

proj-community
'''

def get_yaml_dict(yaml_file):
	"""Return a yaml_dict from reading yaml_file. If yaml_file is empty or
	doesn't exist, return an empty dict instead."""
	try:
		with open(yaml_file, "r") as file_:
			yaml_dict = yaml.safe_load(file_.read()) or {}
		return yaml_dict
	except FileNotFoundError:
		return {}

def format_scheme(scheme):
	"""Change $scheme so it can be applied to a template."""
	scheme["scheme-name"] = scheme.pop("scheme")
	scheme["scheme-author"] = scheme.pop("author")
	bases = ["base{:02X}".format(x) for x in range(0, 24)]
	for base in bases:
		scheme["{}-hex".format(base)] = scheme.pop(base)


def main():
	''' Main entry point for cli '''
	parser = argparse.ArgumentParser(
	description="Generate a table to put in a scheme project readme")
	parser.add_argument("scheme", action="store",
	help="base24 scheme file")

	args = parser.parse_args()
	# Check for and report level8 errors
	if not os.path.isfile(args.scheme):
		print(args.scheme + " is not a valid file")
		sys.exit(1)


	scheme = get_yaml_dict(args.scheme)
	format_scheme(scheme)
	file_content = pystache.render(template, scheme)
	print(file_content)



if __name__ == "__main__":
	main()


windowsTerminal2b24.py
#!/usr/bin/env python3
"""Convert profiles.json to base24 scheme
"""
import os
import sys
import argparse
import yaml
import commentjson
from metprint import (
	LogType,
	Logger
)
import base24tools


def genBase24s(winTerm):
	"""Generate the base24 json object

	Args:
		winTerm (str): a dictionary representing the source theme

	Returns:
		dict[]: base24 dicts to write to scheme file
	"""
	base24s = []
	for scheme in winTerm:
		base24 = {"author": "WinTerm2B24", "scheme": scheme["name"]}
		base24lookup = {
			"base00": "background",
			"base01": "black", #Black
			"base02": "brightBlack", #Bright black
			"base05": "foreground", #Bright black
			"base06": "white", #White
			"base07": "brightWhite", #Bright white
			"base08": "red", #Red
			"base09": "yellow", #Yellow
			"base0A": "brightYellow", #Bright yellow
			"base0B": "green", #Green
			"base0C": "cyan", #Cyan
			"base0D": "blue", #Blue
			"base0E": "purple", #Purple
			"base12": "brightRed", #Bright red
			"base13": "brightYellow", #Bright yellow
			"base14": "brightGreen", #Bright green
			"base15": "brightCyan", #Bright cyan
			"base16": "brightBlue", #Bright blue
			"base17": "brightPurple", #Bright purple
		}

		base24s.append(base24tools.process(base24, base24lookup, scheme, 2, 5))

	return base24s


def winTerm2hex(filename):
	"""Generate the shemes (without #)

	Args:
		filename (str): filename from args

	Returns:
		dict: a dictionary representing the source theme
	"""
	profiles = commentjson.loads(open(filename).read())
	for scheme in profiles["schemes"]:
		for colour in scheme:
			scheme[colour] = scheme[colour].replace("#", "")

	return profiles["schemes"]



def main():
	''' Main entry point for cli '''
	parser = argparse.ArgumentParser(
	description="Convert profiles.json to base24 scheme")
	parser.add_argument("file", action="store",
	help="profiles.json")
	args = parser.parse_args()
	# Check for and report level8 errors
	if not os.path.isfile(args.file):
		Logger().logPrint(args.file + " is not a valid file", LogType.ERROR)
		sys.exit(1)

	filename = args.file


	base24s = genBase24s(winTerm2hex(filename))

	for base24 in base24s:
		with open(base24["scheme"]+".yaml", "w") as outfile:
			Logger().logPrint("writing \"" + base24["scheme"] + "\" to file", LogType.SUCCESS)
			yaml.dump(base24, outfile)

if __name__ == '__main__':
	main()


templateGen.py
"""Base24 2020
Genrate a base24 template from a theme and a scheme
"""
import argparse
import re
import yaml

def get_yaml_dict(yaml_file):
	"""Return a yaml_dict from reading yaml_file. If yaml_file is empty or
	doesn't exist, return an empty dict instead."""
	try:
		with open(yaml_file, "r") as file_:
			yaml_dict = yaml.safe_load(file_.read()) or {}
		return yaml_dict
	except FileNotFoundError:
		return {}

def reverse_hex(hex_str):
	"""Reverse a hex foreground string into its background version."""
	hex_str = "".join([hex_str[i : i + 2] for i in range(0, len(hex_str), 2)][::-1])
	return hex_str


def fuzzReplaceOne(parts_fuzz, replaceWith, outText, mode):
	"""Find and replace for one enumerated colour

	Args:
		parts_fuzz (int[]): red (blue if reversed), green, blue (red if reversed)
		replaceWith (str): the colour id
		outText (str): text to find and replace
		mode (str): colour mode

	Returns:
		str: text with find and replace done
	"""
	if "hex" in mode:
		replaceText = re.compile(re.escape(
			"{:02x}".format(parts_fuzz[0])+"{:02x}".format(parts_fuzz[1])+"{:02x}".format(parts_fuzz[2])
		), re.IGNORECASE)
		outText = replaceText.sub("{{"+replaceWith+"-hex}}", outText)
	elif "dec" in mode:
		# Round and do extras
		part0_dec = parts_fuzz[0]/255
		part1_dec = parts_fuzz[1]/255
		part2_dec = parts_fuzz[2]/255
		for decimalPlaces in range(2, 9):
			replaceText = re.compile(re.escape(
				str(round(part0_dec, decimalPlaces))+","+
				str(round(part1_dec, decimalPlaces))+","+
				str(round(part2_dec, decimalPlaces))
			), re.IGNORECASE)
		outText = replaceText.sub("{{"+replaceWith+"-dec-r}},{{"+
			replaceWith+"-dec-g}},{{"+replaceWith+"-dec-b}}", outText)

		# As with others
		replaceText = re.compile(re.escape(
			str(part0_dec)+","+str(part1_dec)+","+str(part2_dec)
		), re.IGNORECASE)
		outText = replaceText.sub("{{"+replaceWith+"-dec-r}},{{"+
		replaceWith+"-dec-g}},{{"+replaceWith+"-dec-b}}", outText)
	else:
		replaceText = re.compile(re.escape(
			str(parts_fuzz[0])+","+str(parts_fuzz[1])+","+str(parts_fuzz[2])
		), re.IGNORECASE)
		outText = replaceText.sub("{{"+replaceWith+"-rgb-r}},{{"+
		replaceWith+"-rgb-g}},{{"+replaceWith+"-rgb-b}}", outText)

	return outText


def fuzzReplace(schemeDict, key, fuzz, outText, mode):
	"""Replace colour(s) with a single base24 colour id

	Args:
		schemeDict (dict): scheme as a python dictionary
		key (str): colour id
		fuzz (int): +- by
		outText (str): text to do find and replace on
		mode (str): colour mode

	Returns:
		str: text with find and replace done
	"""
	if "reverse" in mode:
		colour = reverse_hex(schemeDict[key])
	else:
		colour = schemeDict[key]
	part0 = int(colour[0:2], 16) - fuzz
	part1 = int(colour[2:4], 16) - fuzz
	part2 = int(colour[4:6], 16) - fuzz

	for part0_fuzz in range(part0, part0 + fuzz * 2 + 1):
		for part1_fuzz in range(part1, part1 + fuzz * 2 + 1):
			for part2_fuzz in range(part2, part2 + fuzz * 2 + 1):
				outText = fuzzReplaceOne([part0_fuzz, part1_fuzz, part2_fuzz], key, outText, mode)

	return outText


def templateGen(scheme, theme, mode="hex", fuzz=0):
	"""Generate a template for a given file

	Args:
		scheme (file): yaml dictionary containing colours
		theme (file): file containing theme colours
		mode (str, optional): colour mode. Defaults to "hex".
		fuzz (int, optional): +- colors for themes that have slight shades.
		Defaults to 0.
	"""
	schemeDict = get_yaml_dict(scheme)

	outText = open(theme, "r").read()
	schemeDict.pop("scheme")
	schemeDict.pop("author")

	for key in schemeDict:
		outText = fuzzReplace(schemeDict, key, fuzz, outText, mode)

	print(outText)


if __name__ == "__main__":
	"""Main method, do argparsing and call templateGen
	"""
	parser = argparse.ArgumentParser("templateGen",
	description="Generate a base24 template from an existing scheme and a theme file")
	parser.add_argument("colour_scheme",
	help="relative or abs path to the base24 colour scheme")
	parser.add_argument("theme",
	help="relative or abs path to the theme file")
	parser.add_argument("--mode", action="store", default="hex",
	help="""color format: hex (ff00aa), reversehex (aa00ff), rgb (255,0,170),
	reversergb (170,0,255), dec (1.0,0,0.666), reversedec (0.666,0,1.0)""")
	parser.add_argument("--fuzz", action="store", default=0,
	help="find 'close' colours and replace these ((r,g,b)+-fuzz: default=0, max-recommended=5)")
	args = parser.parse_args()
	templateGen(args.colour_scheme, args.theme, args.mode, int(args.fuzz))
